### DISCUSSION TITLE
WebOne as MSN Messenger gateway

### DISCUSSION THREAD

**Discussion by atauenis** [atauenis/webone#40] [General]
Found that MSN Messenger 2.2 (WinME), 4.0 (WinXP), 4.6, 4.7 (WinXP SP2), 5.0, 7.0 and some other versions can use [HTTP gateway]( https://gitlab.com/escargot-chat/server/-/wikis/http-gateway) if they can't connect to server directly.

With [Escargot registry tweak](https://escargot.chat/connecting) they are working properly. Escargot's server is still accessible (with some small nuances), so gateways aren't used at all. However seems that theoretically that tweak is not really need, as we can use WebOne to work as the HTTP gateway:
```INI
;webone.conf
[Edit:^http://gateway.messenger.hotmail.com]
AddInternalRedirect=https://m1.escargot.chat/
```
So the connection to now-dead `messenger.hotmail.com` gateway should be do via WebOne, which is redirecting the connection to Escargot. Trying...
```
04.2021 23:25:43	WebOne 0.10.7.0 (Win32NT-64, Runtime 3.1.10) log started.
...
24.04.2021 23:25:49.352+5000	>POST http://gateway.messenger.hotmail.com/gateway/gateway.dll?Action=open&Server=NS&IP=messenger.hotmail.com (192.168.1.69)
24.04.2021 23:25:49.352+810046	 Fix to https://m1.escargot.chat/ internally
24.04.2021 23:25:49.352+845048	>Uploading 0K of application/x-msn-messenger...
24.04.2021 23:25:49.352+19081091	 200 OK. Body 0K of application/octet-stream [Binary].
24.04.2021 23:25:49.352+19096092	<Done.
```
All seems to be okay, but MSN saying that no MSN at all. :)
![nomsn](https://user-images.githubusercontent.com/4800266/115972254-f5d34900-a555-11eb-89af-d92e2a3465d6.png)

If try MSN 7.0, we're see 2 attempts to login, but same error. This is what MSN sends & receives (WebOne traffic dump):

```
POST request to http://m1.escargot.log1p.xyz/gateway/gateway.dll?Action=open&Server=NS&IP=messenger.hotmail.com HTTP/1.1
Proxy-Connection: Keep-Alive
Pragma: no-cache
Content-Length: 27
Content-Type: text/xml; charset=utf-8
Accept: */*
Host: m1.escargot.log1p.xyz
REAL-User-Agent: Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; MSN Messenger 7.0.0820)
SENT-User-Agent: Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; MSN Messenger 7.0.0820) WebOne/0.10.7.0
upgrade-insecure-requests: 1
sec-fetch-mode: navigate
sec-fetch-site: same-origin
sec-fetch-user: ?1
Body goes below. CAUTION: Private area!
VER 12 MSNP11 MSNP10 CVR0



Response 200 OK HTTP/1.0
Access-Control-Allow-Origin: *
Access-Control-Expose-Headers: X-MSN-Messenger
Date: Sun, 25 Apr 2021 09:36:11 GMT
Server: Caddy, Python/3.6, aiohttp/2.3.3
X-Msn-Messenger: SessionID=4ebbd951-3cdf-43aa-9ac5-98fa52140bdc; GW-IP=messenger.hotmail.com
Via: HTTP/1.0 WebOne/0.10.7.0
Content-Type: application/x-msn-messenger
Transfer-Encoding: chunked
Complete.

---------

POST request to http://m1.escargot.log1p.xyz/gateway/gateway.dll?Action=open&Server=NS&IP=messenger.hotmail.com HTTP/1.1
Proxy-Connection: Keep-Alive
Pragma: no-cache
Content-Length: 0
Accept: */*
Host: m1.escargot.log1p.xyz
REAL-User-Agent: Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; MSN Messenger 7.0.0820)
SENT-User-Agent: Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; MSN Messenger 7.0.0820) WebOne/0.10.7.0
upgrade-insecure-requests: 1
sec-fetch-mode: navigate
sec-fetch-site: same-origin
sec-fetch-user: ?1
No body.

Response 200 OK HTTP/1.0
Access-Control-Allow-Origin: *
Access-Control-Expose-Headers: X-MSN-Messenger
Date: Sun, 25 Apr 2021 09:36:12 GMT
Server: Caddy, Python/3.6, aiohttp/2.3.3
X-Msn-Messenger: SessionID=c8eefb0d-037d-42ca-aa8e-51da63def664; GW-IP=messenger.hotmail.com
Via: HTTP/1.0 WebOne/0.10.7.0
Content-Type: application/x-msn-messenger
Transfer-Encoding: chunked
Complete.
```

As I can think from this (note that I am used MSN almost first time in my life :) ), Messenger is sending its version to server, then gots an session ID, and nothing more. Sadly, but I don't know how log should be appear when the `gateway.messenger.hotmail.com` was worked. Probably it should send login command through this gateway, but somewhy it doesn't.

The MSN gateway support may be a good thing, because MSN Messenger 4.7+ on Win9x/XP/Vista don't like HTTPS configuration of Escargot servers, and a client patch is available only for or WinXP and newer. So a HTTP-gateway is a single way to connect from unpatched systems.

**Comment by atauenis** [comment]
Read [here](https://gitlab.com/escargot-chat/server/-/issues/13) an idea. Probably need to edit HTTP header about server location:
```
X-Msn-Messenger: SessionID=4ebbd951-3cdf-43aa-9ac5-98fa52140bdc; GW-IP=messenger.hotmail.com

to

X-Msn-Messenger: SessionID=4ebbd951-3cdf-43aa-9ac5-98fa52140bdc; GW-IP=m1.escargot.log1p.xyz
or
X-Msn-Messenger: SessionID=4ebbd951-3cdf-43aa-9ac5-98fa52140bdc; GW-IP=%ProxyHost%
```

Currently I haven't checked this, will check later.

The other probably reason why WebOne can't act as MSN gateway is that WebOne don't support Keep-Alive connections. It is closing the connection just after end of response data transfer. But I can't find anywhere if the MSN rely on HTTP 1.1 Keep-Alive or if it can work over HTTP 0.9/1.0 proxies like WebOne.

**Reply by atauenis** [reply]
Found an old article about MSNP-over-HTTP gatewaying, and seems that it have not requirement to Keep-Alive. But the header, however, actively used by MSN Messenger.
http://www.hypothetic.org/docs/msn/general/http_connections.php
http://www.hypothetic.org/docs/msn/general/http_examples.php

GW-IP parameter is not really important.

**Comment by pixaline** [comment]
There is information about the gateway protocol here:
https://wiki.nina.chat/wiki/Protocols/MSNP/Gateway

And some general information of how MSN communicates:
https://wiki.nina.chat/wiki/Protocols/MSNP/Tweener
https://wiki.nina.chat/wiki/Protocols/MSNP/Commands/USR

I have been trying to get Messenger for Mac to work with both escargot and my own server.
So far it has failed because it expects SSL 2.0.

**Reply by atauenis** [reply]
I've asked on Escargot GitLab, and a developer answered that HTTP Gateway mode support in Escargot is not complete.
https://gitlab.com/escargot-chat/server/-/issues/113#note_566889096

**Reply by Animadoria** [reply]
Hi, an *actual* Escargot developer here (that person is not a developer). It should be working properly, as far as I know.

Old SSL versions are not supported due to the (horrible) HTTP software we use (Caddy). It auto generated an HTTPS certificate that only allows TLSv1.0+.

(We're working on it, keep an eye out)

**Comment by atauenis** [comment]
With WebOne v0.12.2 release, which added Keep-Alive support, MSN starts working!

The only is what need is the Edit Set in a WebOne configuration file. Call it `msn.conf`, for example:
```INI
[Edit:(^http://.*messenger\.hotmail.*\.com/)(.*)]
AddInternalRedirect=https://msnmsgr.escargot.chat/$2
```

Then set the WebOne address&port in MSN Messenger options:
![MSN](https://user-images.githubusercontent.com/4800266/188328309-8e4dc7e5-a030-48fb-a893-d3f31db4eb08.png)

And unpatched Windows Messenger 4.0 from Windows XP RTM connects to Escargot server.

**Comment by atauenis** [comment]
Good news again. In new 0.16 release there is enough well [support of HTTPS proxying](https://github.com/atauenis/webone/wiki/HTTPS-and-SSL), including the `CONNECT` method used to access SSL/HTTPS servers through proxy servers. Previous WebOne releases ignored this method even without any messages in log.

Now even WLM 2009 is working through Escargot and WebOne. It is not required to use patched versions (however patched are also working). Original MSN/Live Messenger signed by Microsoft also can connect to Escargot servers when the proxy server is configured to redirect HTTPS. Here is the edit set, tested with original builds of MSN 3.6.0039, 6.2.0208 7.0.0820, 7.5.0324, WLM 8.1.0178, 8.5.1302, 14.0.8117:
```INI
; Windows Live Messenger - Escargot gateway configuration.
; Don't forget to import WebOne Root Certificate in Windows registry.

; 		Login
[Edit:^http://messenger.hotmail.com/(.*)]
AddInternalRedirect=https://msnmsgr.escargot.chat/$1

; 		Gateway
[Edit:^http://gateway.messenger.hotmail.com/(.*)]
AddInternalRedirect=https://msnmsgr.escargot.chat/$1

; MSN 5 - 7.0	Login
[Edit:(^http://nexus.passport.com/rdr/pprdr.asp)]
AddInternalRedirect=https://m1.escargot.chat/nexus-mock

; MSN 7.5	Login
[Edit:^http://loginnet.passport.com/(.*)]
AddInternalRedirect=https://msnmsgr.escargot.chat/$1

; MSN 8		Login
[Edit:^http://login.live.com/(.*)]
AddInternalRedirect=https://msnmsgr.escargot.chat/$1

; MSN 6+	configuration
[Edit:^http://config.messenger.msn.com/(.*)]
AddInternalRedirect=https://conf.escargot.chat/$1

; MSN 8+	contacts
[Edit:^http://byrdr.omega.contacts.msn.com/(.*)]
AddInternalRedirect=https://msnmsgr.escargot.chat/$1

; 		Offline messaging
[Edit:^http://ows.messenger.msn.com/(.*)]
AddInternalRedirect=https://msnmsgr.escargot.chat/$1
[Edit:^http://rsi.hotmail.com/(.*)]
AddInternalRedirect=https://msnmsgr.escargot.chat/$1

; MSN 8+	Windows Live Space
[Edit:^http://tkrdr.storage.msn.com/(.*)]
AddInternalRedirect=https://msnmsgr.escargot.chat/$1

; MSN 9+	What's New service
[Edit:^http://sup.live.com/(.*)]
AddInternalRedirect=https://msnmsgr.escargot.chat/$1

[Edit:http://login.live.com/login2.srf(.*)]
AddInternalRedirect=https://msnmsgr.escargot.chat/login

; MSN 7 (old Escargot patch)
[Edit:http://m1.escargot.chat/login]
AddInternalRedirect=https://msnmsgr.escargot.chat/login

[Edit:^http://status.messenger.msn.com/(.*)]
AddInternalRedirect=http://web.archive.org/web/20090212104204id_/http://status.messenger.msn.com/$1
```
Versions 4.7 of the Messenger and up are requiring Windows XP SP3 at this moment to work properly over WebOne. This is caused because they are using HTTPS to authorize, and WebOne is generating a SHA256 certificate for connection. SHA256 support was introduced only since Windows Vista RTM and XP SP3.

**Reply by Animadoria** [reply]
Note: these domains are deprecated and may not work in the future:

```ini
; MSN 8+	contacts
[Edit:^http://byrdr.omega.contacts.msn.com/(.*)]
AddInternalRedirect=https://ebyrdromegactcsmsn.log1p.xyz/$1

; 		Offline messaging
[Edit:^http://ows.messenger.msn.com/(.*)]
AddInternalRedirect=https://eowsmsgrmsn.log1p.xyz/$1
[Edit:^http://rsi.hotmail.com/(.*)]
AddInternalRedirect=https://ersih.log1p.xyz/$1

; MSN 8+	Windows Live Space
[Edit:^http://tkrdr.storage.msn.com/(.*)]
AddInternalRedirect=https://etkrdrstmsn.log1p.xyz/$1
```

Replace them with msnmsgr.escargot.chat/$1 for all cases.

**Reply by atauenis** [reply]
Thanks. It was difficult to google actual URLs for them. Corrected.

**Comment by atauenis** [comment]
With WebOne 0.17, the bug which breaks work of SSL under pre-XPSP3 systems is fixed. And now original MSN Messenger 7.5/8.1 is working under Windows 2000 SP4 and XP RTM (on Win2k requires manual unzipping).
![MSN-7547-Win2k](https://github.com/atauenis/webone/assets/4800266/8a7e606e-d58b-4323-a971-79d581f4cb57)
[MSN Messenger 7.5.zip](https://github.com/atauenis/webone/files/15467813/MSN.Messenger.7.5.zip)


Sadly, but something has changed on Escargot servers, and MSN 4.7/5/6/7.0 (not 7.5 and up) are currently not able to log in (even under WinXP + ProxHTTPSproxyMII, where previously worked). Hope, the bug is temporary.

```
[20:08:03.582]	PPLogin	Attempting to hit Nexus
[20:08:03.582]	PPLogin	Downloading Nexus Data from https://nexus.passport.com/rdr/pprdr.asp
[20:08:03.582]	Net	CHTTPNetworkLayer::SendHttpData (3) >> Posting to URL 
http://messenger.hotmail.com/gateway/gateway.dll?Action=poll&SessionID=7150b3bf-d823-43b1-a58d-8f275b00cb3a
[20:08:04.033]	Net	CHTTPNetwokLayer::HttpCallback (3) >> called back with 200
HTTP/1.0 200 OK
Date: Mon, 27 May 2024 20:08:04 GMT
Access-Control-Allow-Origin: *
Access-Control-Expose-Headers: X-MSN-Messenger
X-MSN-Messenger: SessionID=7150b3bf-d823-43b1-a58d-8f275b00cb3a; GW-IP=messenger.hotmail.com
Server: Windows95
Content-Length: 0
Via: HTTP/1.0 WebOne/0.17.0.0
Content-Type: application/x-msn-messenger
Proxy-Connection: keep-alive
Connection: Keep-Alive


[20:08:04.283]	PPLogin	Nexus Called back with
HTTP/1.0 200 OK
Date: Mon, 27 May 2024 20:08:04 GMT
Passporturls: DALogin=https://m1.escargot.chat/login
Server: Windows95
Access-Control-Expose-Headers: Authentication-Info,X-MSN-Messenger
Access-Control-Allow-Origin: *
Content-Length: 7
Via: HTTP/1.0 WebOne/0.17.0.0
Content-Type: text/plain; charset=utf-8
Proxy-Connection: keep-alive
Connection: Keep-Alive


[20:08:04.283]	PPLogin	CMNSNet::OnPassportLoginCallback > called back with status code 404
[20:08:04.383]	Net	CConnectionManager::OnConnectionStateChange recieved state change for id(3) from 8 to 0
[20:08:04.383]	Net	CConnectionManager::OnConnectionStateChange recieved state change for id(0) from 1 to 0
[20:08:04.383]	Net	CConnectionManager::OnConnectionStateChange IGNORING state change for id(0) from 1 to 0
[20:08:04.383]	Net	CSocksNetworkLayer::OnConnectionStateChange (0) >> changing from 1 to 1
[20:08:04.383]	Net	CConnectionManager::OnConnectionStateChange recieved state change for id(0) from 0 to 1
[20:08:04.383]	Net	CConnectionManager::OnConnectionStateChange IGNORING state change for id(0) from 0 to 1
[20:08:04.383]	Net	CConnectionManager::OnConnectionStateChange recieved state change for id(1) from 1 to 0
[20:08:04.383]	Net	CConnectionManager::OnConnectionStateChange IGNORING state change for id(1) from 1 to 0
[20:08:04.383]	Net	CSocksNetworkLayer::OnConnectionStateChange (1) >> changing from 1 to 1
[20:08:04.383]	Net	CConnectionManager::OnConnectionStateChange recieved state change for id(1) from 0 to 1
[20:08:04.383]	Net	CConnectionManager::OnConnectionStateChange IGNORING state change for id(1) from 0 to 1
[20:08:04.383]	Net	CConnectionManager::OnConnectionStateChange recieved state change for id(2) from 1 to 1
[20:08:04.383]	Net	CConnectionManager::OnConnectionStateChange IGNORING state change for id(2) from 1 to 1
[20:08:04.393]	Net	CConnectionManager::OnConnectionStateChange recieved state change for id(3) from 0 to 1
```

When it will be fixed, probably, all versions of MSN Messenger (unpatched editions) would begin work under Windows 9x/Me/2000/NT4. Including 4.7-7.0.

**Reply by Animadoria** [reply]
Thank you for your work! Question -- is that install of MSN Messenger a pre-patched version or an unpatched version?

**Reply by atauenis** [reply]
It's unpatched version. Just a copy of `C:\Program Files\MSN Messenger\` from WinXP (7.5's installer is checking for OS version, but actually 7.5 is compatible with Win2000).

All patched versions are requiring XP SP2+ to run.

**Reply by Animadoria** [reply]
Oh! That makes sense. Then we need to change "Passporturls" to "PassportURLs". The original, official one was "PassportURLs", but we previously (not anymore) used Caddy which enforced that naming convention of "Passporturls".

We have two options for this problem:
- Check for the "Via" header, check if it's via WebOne and change it to PassportURLs in the server [edit: seems that Via is sent back to the client on the response, not the other way around]
- Make WebOne send the correct header on the proxy.

This problem will be fixed on Escargot 2.0, basically a full rewrite, but for now we have to come up with a solution. I'm more than happy to make a workaround on the server.

**Reply by Animadoria** [reply]
I just pushed an update to the server that should fix the issue. I check if "WebOne" is in the User-Agent, and if it is, I send the correct header. This should theoretically fix it, hopefully. Keep me posted.

**Reply by atauenis** [reply]
Checked, it's working now. Thanks. Now it's possible to log in even from Windows 98 using MSN 5/6/7.

**Reply by freakedenough** [reply]
Its not working for MSN 5.0, I get the "password service unavailable" error. msn7 on the same pc with same webone config works. unpatched, windows me.

**Reply by atauenis** [reply]
> Its not working for MSN 5.0, I get the "password service unavailable" error. msn7 on the same pc with same webone config works. unpatched, windows me.

Does Internet Explorer works with TLS 1.0 128-bit security via your WebOne installation?

**Reply by carlosbaldo95** [reply]
I've been trying to use WebOne for MSN Messenger in the past few days. However, I just can't figure out how to make it work. The server is running Windows 11 22H2 (with the required regfile installed), the client Windows 98 Second Edition (with IE 6, DUN 1.4 and WebOne cert installed). By looking at the WebOne output, everything is fine, but Messenger always throws an error after trying to login.

Versions 6.2 and 7.0 give the 81000306 error after several minutes. Older versions don't give an error code at all and fail much sooner. Windows Live Messenger 2009 in the server logs into my account just fine. Any ideas?

**Comment by atauenis** [comment]
@carlosbaldo95

> the client Windows 98 Second Edition (with IE 6, DUN 1.4 and WebOne cert installed)

Does IE6 opening https:// websites via WebOne? MSN Messenger 4.7 and up is using same common system things as MSIE using to connect to HTTPS based Windows Live ID/.NET Passport services.

> Older versions don't give an error code at all and fail much sooner.

Have you tried 3.5/3.6/4.0 versions? They're don't use SSL/HTTPS at all.

Also I have just now re-checked WebOne 0.17.6/Win8.1 + MSN 7.0/Win98SE combination, it's working.

**Reply by carlosbaldo95** [reply]
IE 6 browsing in https sites is working. They appear in a broken state because of IE being pretty outdated, but that was expected.
![WebOne](https://github.com/user-attachments/assets/cc57c96f-2cf7-47bc-88db-0410941f5429)

The problem is with Messenger clients. Version 3.6 works fine, WebOne successfully redirects the connection to Escargot servers and logs into my account. Newer versions, even without errors in the server's output, just can't authenticate the same way. I'm sure it's SSL related, but I just don't know how to solve it.

This is the MSN Messenger 7.0 connection log. Maybe it's useful for finding out where the problem is.
[MsnMsgr.txt](https://github.com/user-attachments/files/22861130/MsnMsgr.txt)

**Reply by carlosbaldo95** [reply]
The problem was successfully solved by the CrossTalk team. Just add these two lines to escargot.conf:

[Server:^http://nexus.passport.com/(.*)]
RemoteHttpVersion==1.1

Works perfectly. Hope it helps everyone facing this problem.
<img width="1366" height="721" alt="image (1)" src="https://github.com/user-attachments/assets/c490dedd-29b4-49e0-9eea-cf21d79af5c5" />

**Comment by atauenis** [comment]
Found a way how to reanimate Windows XP's .NET Passport Wizard.

It is connecting to Nexus server, then reads Nexus response (which initially contained [many](https://learn.microsoft.com/en-us/windows/win32/winhttp/passport-authentication-in-winhttp#passport-configuration-overrides-used-for-testing) options, but Escargot replies with only `DALogin`), and opens HTML-based wizard from `DAReg` option. Also it can read from `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Internet Settings\Passport` `RegistrationUrl`, but not always.

So, added a Edit set that adds DAReg option (in escargot.conf of WebOne 0.18.x):
```INI
[Edit:^http://nexus.passport.com/(.*)]
AddInternalRedirect=https://msnmsgr.escargot.chat/$1
AddResponseHeader=PassportURLs: DALogin=https://ds.escargot.nina.chat/login,DAReg=http://%Proxy%/msnreg.htm
```

Then it is need to create `/html/msnreg.htm` file with the HTML-based registration wizard stub. Just a HTML page don't work, Windows verifies it. The file syntax is some like [XP Web Publishing Wizard](https://learn.microsoft.com/en-us/windows/win32/lwef/publishing-wizards) syntax.

Now it's working.

<img width="1024" height="768" alt="passpwiz" src="https://github.com/user-attachments/assets/785f2460-4daf-4948-8eb4-c4a8e35be172" />

---

Probably there are other wizards, located at these URLs:

```
PassportURLs: DARealm=Passport.net,
DALogin=login.passport.com/login2.asp,
DAReg=https://register.passport.com/defaultwiz.asp,
Properties=https://memberservices.passport.com/ppsecure/MSRV_EditProfile.asp,
Privacy=https://www.passport.com/consumer/privacypolicy.asp,
GeneralRedir=https://nexusrdr.passport.com/redir.asp,
Help=https://memberservices.passport.com/UI/MSRV_UI_Help.asp,
ConfigVersion=2

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Internet Settings\Passport]
"LoginServerRealm"="Passport.Net"
"LoginServerUrl"="https://login.live.com/login2.srf"
"RegistrationUrl"="https://login.live.com/err.srf"
"Properties"="https://account.live.com/EditProf.aspx?lcid=%L"
"Privacy"="https://login.live.com/gls.srf?urlID=MSNPrivacyStatement&lc=%L"
"GeneralRedir"="https://nexusrdr.passport.com/redir.asp"
"Help"="https://account.live.com/?lcid=%L&dc=PPRDR_Help"
"ConfigVersion"=dword:0000000f
```

First list is from pprdr.asp example answers, and the second list is from XP/Vista systems with WLM 8 or 2009. **DALogin** is login server, known. **DARealm** is for MSIE/Windows password manager, probably a case-independed constant `"Passport.Net"`. **Properties** is page opening from Edit Profile button in User Accounts control panel on WinXP systems where the passport is correctly registered.

**Reply by atauenis** [reply]
The privacy statement link in the Passport Wizard is also using `pprdr.asp` to get the link from `Privacy` option, so added it. However, this breaks some MSN Messenger versions (at least 7.0). Fixed by additional edit set:

```INI
[Edit:^http://ds.escargot.nina.chat/loginDALogin=https://ds.escargot.nina.chat/login]
AddInternalRedirect=https://ds.escargot.nina.chat/login
```

Profile management button in XP User Accounts control panel (in state when Windows Messenger 4.7 is logged in) is looking to `Properties` option, so added it too. Now it opens real page.

Also updated the Passport wizard placeholder to match Windows XP look
<img width="741" height="541" alt="xpru_register" src="https://github.com/user-attachments/assets/fce2932b-978f-4a22-8a94-5df1fb23a126" />

Everything is currently in `dev` (0.18.2) branch of WebOne repository.
